libmamba/include/mamba/core/env_lockfile.hpp

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2022, QuantStack and Mamba Contributors
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+
+#ifndef MAMBA_CORE_ENVIRONMENT_LOCKFILE_HPP
+#define MAMBA_CORE_ENVIRONMENT_LOCKFILE_HPP
+
+#include <variant>
+#include <string>
+#include <typeindex>
+#include <optional>
+#include <unordered_map>
+
+#include <tl/expected.hpp>
+
+#include "fsutil.hpp"
+#include "package_info.hpp"
+#include "error_handling.hpp"
+
+namespace mamba
+{
+
+    enum class file_parsing_error_code
+    {
+        unknown_failure,     /// Something failed while parsing but we can't identify what.
+        unsuported_version,  /// The version of the file does not matched supported ver.
+        parsing_failure,     /// The content of the file doesnt match the expected format/language
+                             /// structure or constraints.
+        invalid_data,        /// The structure of the data in the file is fine but some fields have
+                             /// invalid values for our purpose.
+    };
+
+    class EnvLockFileError : public mamba_error
```

Comment:
I added a "m_data" member of type `any` to `mamba_error` so that you can store additional information without having to subclass it. Do you think it would make sense to use it to store the parsing error code?

Reply:
I used it to store the YAMLcpp error type_index, but I could store the code there too. This would mean storing a struct instead then.

Reply:
I pushed the change, also improved the interface. Storing extra data using this technique with inheriting error types seems to work well.